CLI admin 'session' command

dominikl opened the pull request:
# What this PR does

Adds a ~~'session'~~ 'open' and 'close' command to the CLI ~~admin~~ session plugin, which enables the creation and removal of user sessions.

See https://trello.com/c/I2eVKJGQ/19-rfe-create-session-for-user-cli

Although I'm not sure if that solves the problem of the 'long running' sessions. According to https://downloads.openmicroscopy.org/omero/5.4.0/api/ome/api/ISession.html#createSessionWithTimeout-ome.system.Principal-long- , if one specifies no timeout value (the plugin will then pass on `0`) the maximum timeToLive will be used. I have no idea what this maximum timeToLive actually is.
 
# Testing this PR

- Use the plugin to create a user session. Check with Insight (copy/paste session ID as username and password), that you can connect to the server. 
- Use the `close` command to close the session again. Verify (e.g. with Insight) that the connection is closed.
- Create another session with a short timeout, e. g. 30s. Check that you can connect using the session ID, wait a bit, check again that you cannot connect with the session id any longer.
- Check `test_sessions.py` integration tests

dominikl committed:
Add 'session' command to CLI admin plugin

Allows to open and close user sessions

joshmoore commented:
Few quick EOW thoughts:

 * :+1: for having this available (especially with all the recent mailing list threads)
 * A bit concerned with having this in `admin` since that usually requires local access. Can I suggest `sessions`?
 * We _may_ want to consider how to make these less official if we need more time to stabilize the API? Unfortunately, decoupling the `sessions` plugin could be difficult since it's required for **all** logins. Long-term this might be functionality that lives under `tokens.py` but we're likely not ready to create that yet. cc: @sbesson 
 * If this is in `sessions`, then `--close` and `--open` (`--create`?) could be sub-commands rather than flags. `open/create` could take an optional username creating for the current user if none is given, similar to the `user password` command.
 * We may want to add a getter for the default timeout values. (Happy to help with that)

mtbc commented:
`sec-primitives.xml`, `SessionBean.java`, `SessionManagerImpl.java` are where it all happens but I think I'd have to do some empirical testing to be sure of behavior.

joshmoore commented:
:smile: yes plus:

 * Default config properties in https://github.com/openmicroscopy/openmicroscopy/blob/develop/etc/omero.properties#L628
 * Mapping of injected properites in https://github.com/openmicroscopy/openmicroscopy/blob/develop/components/server/src/ome/services/sessions/SessionManagerImpl.java#L170

dominikl commented:
So would it be better for now to have that as separate plugin, not in the 'core' `OmeroPy`? If so, could a custom `sessions` plugin extend the 'core' `sessions` plugin? What would actually happen, if I have my own version of the `sessions.py` in the `PYTHONPATH`, would that take priority over the `OmeroPy` one?

joshmoore commented:
At the moment, the ordering is dependent on what order things get read in:

https://github.com/openmicroscopy/openmicroscopy/blob/develop/components/tools/OmeroPy/src/omero/cli.py#L1481

```
        paths = set(self._plugin_paths)
        for x in sys.path:
            x = path(x)
            if x.isdir():
                x = x / "omero" / "plugins"
                if x.exists():
                    paths.add(x)
```
with the *last* plugin winning. This brings up an interesting point that we should clarify. We can't yet make use of it though until that is cleared up. I'd suggest refactoring this PR to use `sessions.py` first, and then let's discuss how comfortable we are with the existing names/arguments. If no one objects, we can include it here. Otherwise, we'll look for a mechanism to put it elsewhere.

dominikl committed:
Move sessions command to session.py plugin

dominikl commented:
Moved the code to the `sessions.py` plugin. Btw, had to use the name `open` instead of `create` because `create` is already used for the login arguments:
`-C, --create                      Create a new session regardless of existing ones`

Any better ideas than `open` for the 'create session' command welcome!

joshmoore commented on components/tools/OmeroPy/src/omero/plugins/sessions.py:
```diff
@@ -631,7 +675,7 @@ def timeout(self, args):
             obj = svc.getSession(uuid)
         except:
-            self.ctx.dbg(traceback.format_exc())
+            self.ctx.dbg(traceback.forhmat_exc())
```
This looks surprising.

joshmoore commented:
* :+1: for the naming and location. No need to use `create`.
 * I'm a bit worried about two `nargs=?` since it can be unclear what's intended. Perhaps use add_user_and_group_arguments: https://github.com/openmicroscopy/openmicroscopy/blob/develop/components/tools/OmeroPy/src/omero/cli.py#L2412
 * We'll probably need some examples of setting a (realistic) timeout.
 * Otherwise, let's let people test and see what they think.

joshmoore commented:
@dominikl : with your next set of changes, would you mind rebasing this onto `develop` and changing the target branch?

dominikl committed:
Use add_user_and_group_arguments()

joshmoore commented on components/tools/OmeroPy/src/omero/plugins/sessions.py:
```diff
@@ -264,6 +275,73 @@ def _configure_dir(self, parser):
     def help(self, args):
         self.ctx.out(LONGHELP % {"prog": args.prog})
 
+    def open(self, args):
```
I think the find_user and find_group methods do this work for you:

https://github.com/openmicroscopy/openmicroscopy/blob/develop/components/tools/OmeroPy/src/omero/cli.py#L2445

dominikl replied:
Thanks. I thought there should already be a method like this somewhere. One of the reasons why I don't like Python (or Javascript), because you usually have to dig through files with thousands of lines of code to find a particular functionality :-/

mtbc commented:
Maybe I am not understanding this properly?

```
(omero) mtbc@ls28101:~/src/openmicroscopy$ omero sessions open --user-name user-2 
Using session d2f83e7c-5c87-4b82-bbc1-4bf6a49580aa (user-2@localhost:4064). Idle timeout: 10 min. Current group: private-1
Traceback (most recent call last):
  File "/home/mtbc/src/openmicroscopy/dist/bin/omero", line 125, in <module>
    rv = omero.cli.argv()
  File "/home/mtbc/src/openmicroscopy/dist/lib/python/omero/cli.py", line 1615, in argv
    cli.invoke(args[1:])
  File "/home/mtbc/src/openmicroscopy/dist/lib/python/omero/cli.py", line 1090, in invoke
    stop = self.onecmd(line, previous_args)
  File "/home/mtbc/src/openmicroscopy/dist/lib/python/omero/cli.py", line 1167, in onecmd
    self.execute(line, previous_args)
  File "/home/mtbc/src/openmicroscopy/dist/lib/python/omero/cli.py", line 1249, in execute
    args.func(args)
  File "/home/mtbc/src/openmicroscopy/dist/lib/python/omero/plugins/sessions.py", line 324, in open
    * 1000))
  File "/home/mtbc/src/openmicroscopy/dist/lib/python/omero_api_ISession_ice.py", line 406, in createSessionWithTimeout
    return _M_omero.api.ISession._op_createSessionWithTimeout.invoke(self, ((p, timeToLiveMilliseconds), _ctx))
omero.SecurityViolation: exception ::omero::SecurityViolation
{
    serverStackTrace = ome.conditions.SecurityViolation: Group sudo is not permitted for group private-1 (gid=3)
	at ome.services.sessions.SessionManagerImpl.validateSessionInputs(SessionManagerImpl.java:938)
	at ome.services.sessions.SessionManagerImpl.access$000(SessionManagerImpl.java:97)
	at ome.services.sessions.SessionManagerImpl$2.doWork(SessionManagerImpl.java:370)
	at sun.reflect.GeneratedMethodAccessor327.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at ome.services.util.Executor$Impl$Interceptor.invoke(Executor.java:562)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.security.basic.EventHandler.invoke(EventHandler.java:153)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.orm.hibernate3.HibernateInterceptor.invoke(HibernateInterceptor.java:111)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy72.doWork(Unknown Source)
	at ome.services.util.Executor$Impl.execute(Executor.java:443)
	at ome.services.sessions.SessionManagerImpl.createSession(SessionManagerImpl.java:364)
	at ome.services.sessions.SessionManagerImpl.createFromRequest(SessionManagerImpl.java:288)
	at ome.services.sessions.SessionBean$2.call(SessionBean.java:167)
	at ome.services.sessions.SessionBean$2.call(SessionBean.java:152)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

    serverExceptionClass = ome.conditions.SecurityViolation
    message = Group sudo is not permitted for group private-1 (gid=3)
}
(omero) mtbc@ls28101:~/src/openmicroscopy$ omero -u user-2 login 
Using session d2f83e7c-5c87-4b82-bbc1-4bf6a49580aa (user-2@localhost:4064). Idle timeout: 10 min. Current group: private-1
(omero) mtbc@ls28101:~/src/openmicroscopy$ 
```

mtbc commented:
Probably will want some coverage in `clitest/`.

dominikl commented:
You need to be logged in as admin (or light admin with 'sudo' permission I guess), although the error message is a bit strange, what's `Group sudo`? Yes, integration test is still missing. Should have probably targeted this PR for 5.4.3.

mtbc commented:
Sounds like you need an `@admin_only` decoration of some kind.

I think earlier I had some confusion because despite being "create a session" it was picking up some existing session and being affected by that, which I suppose makes sense with hindsight.

mtbc commented on components/tools/OmeroPy/src/omero/plugins/sessions.py:
```diff
@@ -247,6 +247,17 @@ def _configure(self, parser):
+        self.add_user_and_group_arguments(open)
+        open.add_argument("--timeout", nargs="?", type=int, default=0,
+                          help="Timeout in seconds (optional; default: "
+                               "maximum possible)")
```
Is it actually creating sessions that last forever? I wonder if it should default instead to whatever the usual is.

dominikl replied:
If one creates a session with no timeout specified, it creates a session with timeout `0` and an idle timeout of `10 min` whatever that means...

dominikl committed:
Use admin_only decorator

joshmoore commented:
Usability RFE:

```
[dist]$ bin/omero sessions open --user-name public_user
4960-9759-1cc6c994b484 (root@xxx:4064). Idle timeout: 10 min. Current group: read-write-1
Session created for user public_user
Session ID: 348bc84d-5d2f-406f-b70c-aaa7919094fd
```

 * stdout could be _only_ the UUID to make this usable from scripts: `UUID=$(bin/omero sessions open...)`
 * include the timeout and timetoidle on stderr (and the group?)
 * less sure, but perhaps an option to save this session in `~/omero/sessions`

dominikl commented:
Where should the session file get stored? 

E.g. logged in as `root`, create session for `user-3`:

Should it go into `˜/omero/sessions/SERVER/root/` or  `˜/omero/sessions/SERVER/user-3/` ?
I tried both. I guess `˜/omero/sessions/SERVER/user-3/` is not particular useful. I thought `˜/omero/sessions/SERVER/root/` would enable the `root` user to list, clear, ... these sessions again. But it doesn't. Calling `˜/omero sessions list` lists only the `root` session and at the same time deletes the `user-3` session file stored in `˜/omero/sessions/SERVER/root/`.

joshmoore commented:
I would assume you would want to make use of `omero.utils.sessions` as the sessions.py plugin does. e.g. https://github.com/openmicroscopy/openmicroscopy/blob/develop/components/tools/OmeroPy/src/omero/plugins/sessions.py#L470

dominikl committed:
Improve output

dominikl committed:
Store session

dominikl committed:
Add integration test for 'sessions open' command

joshmoore commented:
Tested on eel. I was surprised by an interaction including:

 * `omero sessions open --user-name public`
 * `omero sessions list`
 * `omero login public@localhost`

I'm trying to encode as an integration test now

jburel commented:
To consider for 5.4.4 or onhold for now?

joshmoore commented:
Sorry, I'll try to get back to the test. If all goes well, having this in 5.4.4 seems worthwhile, but I wouldn't block on it.

joshmoore committed:
Refactor user methods to cli.py

joshmoore committed:
Fix minor issue in session storage

The primary change here is to pass "username" rather than
"name" returned from "get_current" to the store.add() method.
By passing root's name, trying to login as the new user failed
while trying to login as root logged in as the new user.

joshmoore commented:
The integration test has been extended to cover the issue I saw and a fix put into place.

joshmoore commented:
> less sure, but perhaps an option to save this session in ~/omero/sessions

After having thought about this, I'm even less sure, or if it shouldn't at least be protected by an argument.

joshmoore committed:
Remove listing in share store

joshmoore commented:
After the failure in https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-integration-python/813/testReport/OmeroPy.test.integration.clitest.test_sessions/TestSessions/testOpen/

```
test/integration/clitest/test_sessions.py:348: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
target/omero/cli.py:1093: in invoke
    self.assertRC()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <omero.cli.CLI instance at 0x134a0908>

    def assertRC(self):
        if self.rv != 0:
>           raise NonZeroReturnCode(self.rv, "assert failed")
E           NonZeroReturnCode: assert failed

target/omero/cli.py:1082: NonZeroReturnCode
Standard Error
2018-02-19 10:38:14,202 INFO  [                    omero.util.Resources] (Thread-931) Halted
2018-02-19 10:38:19,298 INFO  [                    omero.util.Resources] (Thread-932) Starting
2018-02-19 10:38:19,314 INFO  [                    omero.util.Resources] (Thread-932) Halted
2018-02-19 10:38:19,369 INFO  [                    omero.util.Resources] (Thread-933) Starting
2018-02-19 10:38:19,484 INFO  [                    omero.util.Resources] (Thread-933) Halted
stdin is not a terminal: cannot request password
```

I rolled back the store methods as well as the third part of the test. `open` semantics *won't* update `~/omero/sessions` unless not until we decide they will (e.g. `--store` or `--login`)

Note: I tried to reproduce this failure locally and could not, even when running the tests behind `ssh`.

jburel commented on components/tools/OmeroPy/test/integration/clitest/test_sessions.py:
```diff
@@ -305,3 +312,33 @@ def testWho(self, who):
+
+    # open session
+    # =======================================================================
+    def testOpen(self, capsys):
```
test_open

jburel commented on components/tools/OmeroPy/test/integration/clitest/test_sessions.py:
```diff
@@ -305,3 +312,33 @@ def testWho(self, who):
+    # =======================================================================
+    def testOpen(self, capsys):
+
+        asUser = self.new_user()
```
better to start using as_user etc.

jburel commented:
https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-integration-python/814/testReport/OmeroPy.test.integration.clitest.test_sessions/TestSessions/testOpen/ green today

jburel commented:
As root

running against a local server or eel:

* ``bin/omero sessions list``. Nothing happens. (Ctrl-C to stop after a period of time)

eel:

* ``bin/omero sessions open --user-name user-1`` works fine
* instead of an error (``username = user.omeName.val
AttributeError: 'NoneType' object has no attribute 'omeName'``) it might be better to send a message if the user specified is not valid

jburel commented:
tested manually the ``open`` option for light admin with and without sudo

with sudo: ``sessions open --user-name public_user`` works
without sudo: ``sessions open --user-name public_user`` throws a ``SecurityViolation: Admin restrictions: Sudo`` as expected

The options should be added as tests to avoid the surprise noticed while working on pyramids removal

jburel commented:
Tested manually ``--user-id``, it works. Integration tests should be added for that option too

jburel commented:
manual test:
* Create a session for any user e.g. user-1
* log in in insight
* close the session via cli
* still possible to interact with the data (not expected)

joshmoore commented:
Logic like that in `killSession` will be needed but not for the current `client`. I'll push a commit refactoring part of `killSession` out into `destroySession`.

joshmoore committed:
destroySession: reduce reference count of sessions

Similar to what killSession does for the current session,
destroySession takes a session uuid and reduces the reference
count until the server-side session is destroyed. Note: waiting
processes won't immediately but on their next keep alive they
will receive an ObjectNotExistException.

dominikl commented:
That's true, you can't join the session any longer (i. e. login with session id), once it's closed via cli, but if it is still active in Insight, you can keep using it. But I assume that's more an Ice client issue, isn't it?

dominikl commented:
Argh, should occassionally reload my tabs, before making comments.... :-)

jburel committed:
Add tests

jburel commented:
@joshmoore @dominikl I have added few more tests for the ``open`` commands

jburel commented:
I missed that one.

jburel committed:
Use helper method

jburel committed:
Add test for close option

jburel commented:
Manually tested the ``close`` option following change in c890102
* Create a session for user-1
* Use the session to connect via insight
* Close the session using ``bin/omero sessions close session_id``
* Try to open an image in insight
* It correctly fails

I have also added an integration test for the ``close`` option 4cd32c5

jburel committed:
Fix doc

joshmoore committed:
Refactor UUID checking

joshmoore commented:
Thanks for the tests, @jburel. I pushed a minor refactoring to reduce the copy-n-paste of the UUID pattern.

jburel commented:
@joshmoore wrong user in https://github.com/openmicroscopy/openmicroscopy/pull/5614#issuecomment-367666935

JBurl commented:
Wrong user.


On Thu, Feb 22, 2018 at 7:34 AM Josh Moore <notifications@github.com> wrote:

> Thanks for the tests, @JBurl <https://github.com/jburl>. I pushed a minor
> refactoring to reduce the copy-n-paste of the UUID pattern.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmicroscopy/openmicroscopy/pull/5614#issuecomment-367666935>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIocb6buUp4xSxSI6MgqJV-xlosNcAwlks5tXV7QgaJpZM4Rcb7U>
> .
>
-- 
*Jessica Burley*
PhD Candidate
Department of Economics
University of Toronto

jburel commented:
```
04:43:33 test/integration/clitest/test_sessions.py::TestSessions::test_open PASSED
04:43:39 test/integration/clitest/test_sessions.py::TestSessions::test_open_with_id PASSED
04:43:49 test/integration/clitest/test_sessions.py::TestSessions::test_open_restricted_admin_no_sudo PASSED
04:43:59 test/integration/clitest/test_sessions.py::TestSessions::test_open_restricted_admin_sudo PASSED
04:44:05 test/integration/clitest/test_sessions.py::TestSessions::test_close PASSED
```
All the tests are green

jburel commented:
@mtbc any comments?
Otherwise I will merge it

mtbc commented:
No current blockers from me.

jburel commented:
Thanks all merging
